### PR TITLE
Add Login feature

### LINKED ISSUE #69: Creating Log in feature
Need to add Log in feature
solves #79

### PR DESCRIPTION
**Pull request by pazMenachem**
Added the following:
login feature, login page.

fix #69 
based #74 
next branch will include log out feature plus change to the navigation bar (view will be depended whether user is logged in or not).
NOTE: sign up is not linked atm on purpose.
image
image

### CODE REVIEW

File: supplier/admin.py [hunk 1]
```diff
@@ -4,7 +4,7 @@
 
 class SupplierAdmin(admin.ModelAdmin):
     list_display = ('user_name', 'first_name', 'last_name',
-                    'password', 'business_name',)
+                    'password',  'business_name',)
```
Comment (yftacherzog) [inline supplier/admin.py:7]: This can be reverted. No?

File: templates/buy_together_app/login.html [hunk 2]
```diff
@@ -0,0 +1,59 @@
...
+						Don't have an account? <a href="" class="ml-2">Sign Up</a>
+					</div>
+			
+				</div>
+			</div>
+		</div>
+	</div>
+</body>
+
+</html>
```
Comment (yftacherzog) [inline templates/buy_together_app/login.html:59]: add newline

**Comment by yftacherzog** [comment]
The removal of `buy_together_app/migrations/__init__.py` is related to this PR?

**Comment by pazMenachem** [comment]
> The removal of `buy_together_app/migrations/__init__.py` is related to this PR?

not really, just caught my eye, though it wont be needed...

**Comment by RawanAbuleil** [comment]
great job paz , looks great 👍

**Comment by yftacherzog** [comment]
> > The removal of `buy_together_app/migrations/__init__.py` is related to this PR?
> 
> not really, just caught my eye, though it wont be needed...

If it's not related then better not include it. It only makes the review go slower. It's also not a good practice regardless for several reasons (e.g. suppose you need to roll a commit back, you want that commit to include a single change, rather than several unrelated things).

Given the time constraints, I'd also suggest avoiding nice-to-have and cosmetic changes at this point, as it will slow down your must-haves.

File: static/css/login_css.css [hunk 3]
```diff
@@ -0,0 +1,63 @@
...
+			box-shadow: none !important;
+			outline: 0px !important;
+		}
+
+		#messages{
+			background-color: grey;
+			color: #fff;
+			padding: 10px;
+			margin-top: 10px;
+		}
```
Comment (yftacherzog) [inline static/css/login_css.css:63]: newline

File: buy_together_app/views.py [hunk 4]
```diff
@@ -7,3 +9,16 @@ def main_page(request):
 
 def description_page(request):
     return render(request, 'buy_together_app/description.html')
+
+
+def log_in_page(request):
```
Comment (yftacherzog) [inline buy_together_app/views.py:14]: We need to have this tested as well. It's a bit more complex than other test cases you did up until now, as it requires calling some django methods. You can use monkeypatch to mock some of those methods and make the test suite use that mock. This way, you don't need to rely on everything being in place, and you can also verify that everything was called as you expected.

I'm going to provide a simplistic approach to go about it. It could be done in a much better way, but for now it should do.

The basics:

```
from unittest.mock import Mock

mock = Mock()
```
This creates a mock object that you can make it take and return whatever you want

```
monkeypatch.setattr(foo, foo.object.__name__, mock)
```
This makes the object called `mock` replace the object called `foo.object.__name__` in module `foo`.

More specifically, suppose that this is the module you're testing (called foo.py):

```
from django.shortcuts import render, redirect
from django.contrib.auth import authenticate, login
from django.contrib import messages


def log_in_page(request):
    if request.method == 'POST':
        user_name = request.POST.get('username')
        user_pass = request.POST.get('password')
        user = authenticate(request, username=user_name, password=user_pass)
        if user is not None:
            login(request, user)
            return redirect('Main Page')
        else:
            messages.info(request, 'User name OR Password is incorrect')
    return render(request, 'buy_together_app/login.html')
```

This is how your tests will look like:
```
import pytest
from unittest.mock import Mock
import foo


class Test_Login:
    @pytest.fixture
    def authenticate_mock(self, monkeypatch):
        mock = Mock(return_value=Mock())
        monkeypatch.setattr(foo, foo.authenticate.__name__, mock)
        return mock

    @pytest.fixture
    def login_mock(self, monkeypatch):
        mock = Mock()
        monkeypatch.setattr(foo, foo.login.__name__, mock)
        return mock

    @pytest.fixture
    def redirect_mock(self, monkeypatch):
        mock = Mock()
        monkeypatch.setattr(foo, foo.redirect.__name__, mock)
        return mock

    @pytest.fixture
    def messages_mock(self, monkeypatch):
        mock = Mock()
        monkeypatch.setattr(foo, "messages", mock)
        return mock

    @pytest.fixture
    def render_mock(self, monkeypatch):
        mock = Mock()
        monkeypatch.setattr(foo, foo.render.__name__, mock)
        return mock

    def test_login_succuss(
        self, authenticate_mock, login_mock, redirect_mock, messages_mock, render_mock
    ):
        request_mock = Mock()
        request_mock.method = "POST"
        output = foo.log_in_page(request_mock)
        assert output == redirect_mock.return_value
        assert authenticate_mock.call_count == 1
        assert login_mock.call_count == 1
        assert redirect_mock.call_count == 1
        assert messages_mock.info.call_count == 0
        assert render_mock.call_count == 0

    def test_login_not_post(
        self, authenticate_mock, login_mock, redirect_mock, messages_mock, render_mock
    ):
        request_mock = Mock()
        request_mock.method = "NOT_POST"
        output = foo.log_in_page(request_mock)
        assert output == render_mock.return_value
        assert authenticate_mock.call_count == 0
        assert login_mock.call_count == 0
        assert redirect_mock.call_count == 0
        assert messages_mock.info.call_count == 0
        assert render_mock.call_count == 1

    def test_login_failed_auth(
        self, authenticate_mock, login_mock, redirect_mock, messages_mock, render_mock
    ):
        request_mock = Mock()
        request_mock.method = "POST"
        authenticate_mock.return_value = None
        output = foo.log_in_page(request_mock)
        assert output == render_mock.return_value
        assert authenticate_mock.call_count == 1
        assert login_mock.call_count == 0
        assert redirect_mock.call_count == 0
        assert messages_mock.info.call_count == 1
        assert render_mock.call_count == 1

```

**Comment by yftacherzog** [comment]
https://github.com/redhat-beyond/BuyTogether/pull/79/commits/ee2ff3ca74e7811968a251402ea03d2211b895df#r1061240096

I've added a comment for how I see the test should look like here. I keep a link here so it's visible also after that comment is resolved.

**Comment by pazMenachem** [comment]
> > > The removal of `buy_together_app/migrations/__init__.py` is related to this PR?
> > 
> > 
> > not really, just caught my eye, though it wont be needed...
> 
> If it's not related then better not include it. It only makes the review go slower. It's also not a good practice regardless for several reasons (e.g. suppose you need to roll a commit back, you want that commit to include a single change, rather than several unrelated things).
> 
> Given the time constraints, I'd also suggest avoiding nice-to-have and cosmetic changes at this point, as it will slow down your must-haves.

yea after i deleted it i thought why not making a PR deleting all the unnecessary files... but as you say its not the right time for cosmetic changes...

**Comment by thetaxas2468** [comment]
> > The removal of `buy_together_app/migrations/__init__.py` is related to this PR?
> 
> not really, just caught my eye, though it wont be needed...

Please do not delete something that comes from the migration that is already built it,please do not also delete init files since they are created automatically in order to prevent any problems in testings

**Comment by thetaxas2468** [comment]
If you can use LogInForm for example it will be easier i guess.

File: buy_together_app/views.py [hunk 5]
```diff
@@ -7,3 +9,16 @@ def main_page(request):
...
+    if request.method == 'POST':
+        user_name = request.POST.get('username')
+        user_pass = request.POST.get('password')
+        user = authenticate(request, username=user_name, password=user_pass)
+        if user is not None:
+            login(request, user)
+            return redirect('Main Page')
+        else:
+            messages.info(request, 'User name OR Password is incorrect')
```
Comment (thetaxas2468) [inline buy_together_app/views.py:23]: There is an option that the username does not exist

Comment (pazMenachem) [reply buy_together_app/views.py:23]: if so the authenticate function will fail and user will be none, will be redirected back to login page.

Comment (thetaxas2468) [reply buy_together_app/views.py:23]: If the username is not existed we want to still in the same page and to tell the user that this username doesnot exist,look at my pr i did something similar

**Comment by pazMenachem** [comment]
> If you can use LogInForm for example it will be easier i guess.

it was easy enough using the html login from the internet👍

**Comment by thetaxas2468** [comment]
> > If you can use LogInForm for example it will be easier i guess.
> 
> it was easy enough using the html login from the internet👍

It is not hard to change it ,it will be quick but i said something that you can or can not do,your choice

File: buy_together_app/views.py (hunk 4 above)
Comment (pazMenachem) [reply buy_together_app/views.py:14]: thanks! did saw some different ways to test the view functions, will use them next testing.
thou the code was REALLY helpful to understand how to test.

File: buy_together_app/views.py (hunk 5 above)
Comment (pazMenachem) [reply buy_together_app/views.py:23]: `            messages.info(request, 'User name OR Password is incorrect')
    return render(request, 'buy_together_app/login.html')`
does exactly that🤷‍♂️

File: buy_together_app/tests/test_login.py [hunk 6]
```diff
@@ -0,0 +1,75 @@
+import pytest
+from unittest.mock import Mock
+from buy_together_app import views
+
+
+class Test_Login:
+    @pytest.fixture
+    def authenticate_mock(self, monkeypatch):
+        mock = Mock()
```
Comment (Omeramsc) [inline buy_together_app/tests/test_login.py:9]: Nice use of mocking. :) check out [create_autospec](https://docs.python.org/3/library/unittest.mock.html#unittest.mock.create_autospec), it is more useful.

File: buy_together_app/views.py [hunk 7]
```diff
@@ -7,3 +9,16 @@ def main_page(request):
 
 def description_page(request):
     return render(request, 'buy_together_app/description.html')
+
+
+def log_in_page(request):
+    if request.method == 'POST':
+        user_name = request.POST.get('username')
+        user_pass = request.POST.get('password')
```
Comment (Omeramsc) [inline buy_together_app/views.py:17]: do note I asked @thetaxas2468 to add encryption to the passwords, it may require decryption on your side.

File: buy_together_app/views.py (hunk 5 above)
Comment (Omeramsc) [inline buy_together_app/views.py:23]: what about using a too short/long usertname or password?

File: buy_together_app/tests/test_login.py [hunk 8]
```diff
@@ -0,0 +1,75 @@
...
+    def test_login_failed_auth(
+        self, authenticate_mock, login_mock, redirect_mock, messages_mock, render_mock
+    ):
+        request_mock = Mock()
+        request_mock.method = "POST"
+        authenticate_mock.return_value = None
+        output = views.log_in_page(request_mock)
+        assert output == render_mock.return_value
+        assert authenticate_mock.call_count == 1
+        assert login_mock.call_count == 0
```
Comment (Omeramsc) [inline buy_together_app/tests/test_login.py:72]: you also have `assert_not_called` for that.

File: buy_together_app/tests/test_login.py [hunk 9]
```diff
@@ -0,0 +1,75 @@
...
+
+    def test_login_failed_auth(
+        self, authenticate_mock, login_mock, redirect_mock, messages_mock, render_mock
+    ):
+        request_mock = Mock()
+        request_mock.method = "POST"
+        authenticate_mock.return_value = None
+        output = views.log_in_page(request_mock)
+        assert output == render_mock.return_value
+        assert authenticate_mock.call_count == 1
```
Comment (Omeramsc) [inline buy_together_app/tests/test_login.py:71]: you also have `assert_called_once` or `assert_called_once_with` for that

File: buy_together_app/tests/test_login.py [hunk 10]
```diff
@@ -0,0 +1,75 @@
...
+        assert render_mock.call_count == 1
+
+    def test_login_failed_auth(
+        self, authenticate_mock, login_mock, redirect_mock, messages_mock, render_mock
+    ):
+        request_mock = Mock()
+        request_mock.method = "POST"
+        authenticate_mock.return_value = None
+        output = views.log_in_page(request_mock)
+        assert output == render_mock.return_value
```
Comment (Omeramsc) [inline buy_together_app/tests/test_login.py:70]: Isn't there a way to check it actually returns the render we expect it to, instead of a meaningless mock?
also we should test we get the `messages.info(request, 'User name OR Password is incorrect')`

File: buy_together_app/tests/test_login.py [hunk 11]
```diff
@@ -0,0 +1,75 @@
...
+        assert authenticate_mock.call_count == 0
+        assert login_mock.call_count == 0
+        assert redirect_mock.call_count == 0
+        assert messages_mock.info.call_count == 0
+        assert render_mock.call_count == 1
+
+    def test_login_failed_auth(
+        self, authenticate_mock, login_mock, redirect_mock, messages_mock, render_mock
+    ):
+        request_mock = Mock()
```
Comment (Omeramsc) [inline buy_together_app/tests/test_login.py:66]: why is that not a fixture?

File: buy_together_app/views.py (hunk 4 above)
Comment (yftacherzog) [reply buy_together_app/views.py:14]: Glad it helped!

File: buy_together_app/tests/test_login.py (hunk 9 above)
Comment (yftacherzog) [reply buy_together_app/tests/test_login.py:71]: @pazMenachem just a word of caution. To keep things simple, the code I suggested was creating mocks directly with `Mock()`. A better approach, as Omer mentioned, is to use `create_autospec` which takes an object as specification and creates a mock containing only the functions the spec has (so calling functions which doesn't exist will fail - as with regular objects).

When not autospecing, your mock will allow calling any function, and asserting on that call will always pass.

So, suppose that you want to use `mock.assert_called_once()` if you happen to misspell it, then your test will always pass regardless of whether the function was called or not.

Bottom line, if you want to use the mock's functions, e.g. `assert_called_once()` (very useful), then only do it when the mock has a spec, and not in the case of creating a blank mock like we did here.

Comment (pazMenachem) [reply buy_together_app/tests/test_login.py:71]: im sure there are better ways, but since we are low on time the code just suppose to do the trick.
since i haven't done any testing of this kind i will learn how to do so in the weekend, but for now i just wanted this part to be merged.

File: buy_together_app/views.py (hunk 7 above)
Comment (pazMenachem) [reply buy_together_app/views.py:17]: the built in user do hashing to the password do no need to encrypt and decrypt

File: buy_together_app/tests/test_login.py [hunk 12]
```diff
@@ -0,0 +1,95 @@
+import pytest
+from unittest.mock import Mock
+from buy_together_app import views
+from django.shortcuts import render, redirect
+from django.test import RequestFactory
+
+
```
Comment (yftacherzog) [inline buy_together_app/tests/test_login.py:7]: @pazMenachem did you decide on keeping all mocks in place eventually? On our call you talked about keeping the mock only for the `messages` call.

Also, you opted towards using the assertion methods (e.g. `assert_called_once`) while still using non-speced mock (e.g. not one which you create with create_autospec). I demonstrated on the call yesterday how this could be dangerous (e.g. if you misspell the function name, it will never fail).

Can you summarize your decisions in regards to these tests?

File: buy_together_app/tests/test_login.py [hunk 13]
```diff
@@ -0,0 +1,95 @@
...
+@pytest.fixture
+def render_mock(monkeypatch, get_request):
+    mock = Mock()
+    monkeypatch.setattr(views, views.render.__name__, mock)
+    mock.return_value = render(get_request, 'buy_together_app/login.html')
+    return mock
+
+
+@pytest.fixture
+def post_method_mock():
```
Comment (yftacherzog) [inline buy_together_app/tests/test_login.py:46]: Should be something like: `request_with_post_mock` (and similar logic for the other with non-post), as the object that you're returning is a mock for the request rather than a mock for the API method.

**Comment by Omeramsc** [comment]
solve conflicts, then ping me for review

File: buy_together_app/tests/test_login.py (hunk 12 above)
Comment (pazMenachem) [reply buy_together_app/tests/test_login.py:7]: since the login feature uses authenticate, which hashes the password, i didnt know how to pass an hashed password, so i stayed with the mock approach.
i forgot how it could be dangerous, eventually we are talking about typo mistake, if i use autospec, the mock knows its method, but if not, there every method i create considered legit, thou assert_called_once is built in method, im thinking maybe about the messages.info, but that mock is patched, remind me what i am missing?

Comment (yftacherzog) [reply buy_together_app/tests/test_login.py:7]:
``` python
In [21]: str_mock = Mock()

In [22]: str_speced_mock = create_autospec(str, instance=True)

In [23]: str_speced_mock.capitalize.asser_not_called()
---------------------------------------------------------------------------
AttributeError                            Traceback (most recent call last)
Cell In [23], line 1
----> 1 str_speced_mock.capitalize.asser_not_called()

File /usr/lib64/python3.9/unittest/mock.py:630, in NonCallableMock.__getattr__(self, name)
    628 elif self._mock_methods is not None:
    629     if name not in self._mock_methods or name in _all_magics:
--> 630         raise AttributeError("Mock object has no attribute %r" % name)
    631 elif _is_magic(name):
    632     raise AttributeError(name)

AttributeError: Mock object has no attribute 'asser_not_called'

In [25]: str_mock.capitalize.asser_not_called()
Out[25]: <Mock name='mock.capitalize.asser_not_called()' id='140300210951600'>

In [26]: assert str_mock.capitalize.call_count == 0
```

In other words, if you misspell the assertion method then it will not cause an exception and the test will pass. 

Ultimately, best option is to use a spec - the easiest way to go about it is to use create_autospec. It will make sure your mock only mocks capabilities the mocked object actually has.

If you're not using autospec, then better not use the mock's functions (e.g. assert_called_once).

To use create_autospec, you'd need something like this:

``` python
from unittest.mock import create_autospec
@pytest.fixture
def authenticate_mock(monkeypatch):
    mock = create_autospec(views.authenticate)
    monkeypatch.setattr(views, views.authenticate.__name__, mock)
    return mock
```

**Review by yftacherzog — APPROVED** [review]

File: buy_together_app/tests/test_login.py (hunk 12 above)
Comment (pazMenachem) [reply buy_together_app/tests/test_login.py:7]: ok got it now, thanks!

**Review by Omeramsc — APPROVED** [review]
Good job, LGTM
@yftacherzog there is still an open thread you're involved with, feel free to merge if/when it is solved.

**Comment by MeitarRizner** [comment]
Greate work Paz!
LGTM

File: buy_together_app/tests/test_login.py (hunk 12 above)
Comment (yftacherzog) [reply buy_together_app/tests/test_login.py:7]: Here we create a mock object. We provide `views.authenticate` as the specifications according to which the mock should be created (what the mock should imitate):
```python
mock = create_autospec(views.authenticate)
```

Here we implant the `mock` object in-place of the `views.authenticate` function inside the `views` model:
```python
monkeypatch.setattr(views, views.authenticate.__name__, mock)
```

**Comment by yftacherzog** [comment]
> Did you check that the authenticate method works well with the creation of the user with the method create_user because it uses SHA256 as encryption(If you can get authenticated with test_migrations that on the data base for client and supplier)
> 
> if yes LGTM!

@pazMenachem Iron this out with @thetaxas2468 and then we can merge this PR.

File: buy_together_app/tests/test_login.py (hunk 12 above)
Comment (pazMenachem) [reply buy_together_app/tests/test_login.py:7]: got it

**Comment by pazMenachem** [comment]
> Did you check that the authenticate method works well with the creation of the user with the method create_user because it uses SHA256 as encryption(If you can get authenticated with test_migrations that on the data base for client and supplier)

not need to test the authenticate method, we work with built in users so nothing to test.
thou did wanted to try something from the db so in the next PR you can see the authenticate works with user from the db.

**Comment by thetaxas2468** [comment]
> > Did you check that the authenticate method works well with the creation of the user with the method create_user because it uses SHA256 as encryption(If you can get authenticated with test_migrations that on the data base for client and supplier)
> 
> not need to test the authenticate method, we work with built in users so nothing to test. thou did wanted to try something from the db so in the next PR you can see the authenticate works with user from the db.

Did you check that create_user and authenticate method works one with the other well then LGTM if not we need to do a little change in that in order to continue in something that working 100% and not fall into bugs that we hard to tell what are they .

**Review by thetaxas2468 — APPROVED** [review]
LGTM
